File: app/src/main/java/com/stedi/randomimagegenerator/app/view/adapters/GeneratorTypeAdapter.java
package com.stedi.randomimagegenerator.app.view.adapters;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.stedi.randomimagegenerator.app.R;
import com.stedi.randomimagegenerator.app.model.data.GeneratorType;
import com.stedi.randomimagegenerator.app.view.components.GeneratorTypeImageLoader;

import butterknife.BindView;
import butterknife.ButterKnife;

public class GeneratorTypeAdapter extends RecyclerView.Adapter<GeneratorTypeAdapter.ViewHolder> {
    private final GeneratorTypeImageLoader imageLoader;
    private final GeneratorType[] generatorType;
    private final ClickListener listener;
    private final boolean isDeselectAllowed;

    private GeneratorType selectedType;
    private GeneratorType targetType;

    public interface ClickListener {
        void onSelected(@NonNull GeneratorType type);

        void onDeselected();

        void onEditSelected();
    }

    public GeneratorTypeAdapter(
            @NonNull GeneratorTypeImageLoader imageLoader,
            @NonNull GeneratorType[] generatorType, @Nullable GeneratorType selectedType, @Nullable GeneratorType targetType,
            @NonNull ClickListener listener, boolean isDeselectAllowed) {
        setHasStableIds(true);
        this.imageLoader = imageLoader;
        this.generatorType = generatorType;
        this.selectedType = selectedType;
        this.targetType = targetType;
        this.listener = listener;
        this.isDeselectAllowed = isDeselectAllowed;
    }

    @Override
    public GeneratorTypeAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.generator_type_item, parent, false));
    }

    @Override
    public void onBindViewHolder(GeneratorTypeAdapter.ViewHolder holder, int position) {
        GeneratorType type = generatorType[position];
        holder.generatorType = type;

        holder.card.setOnClickListener(holder);
        holder.text.setText(type.getStringRes());
        holder.btnEdit.setVisibility(View.INVISIBLE);
        holder.btnEdit.setOnClickListener(holder);
        holder.isSelected.setVisibility(type == selectedType ? View.VISIBLE : View.INVISIBLE);
        holder.image.setImageResource(R.drawable.ic_texture_adapter_rig_image_size);

        imageLoader.load(type, targetType, (params, bitmap) -> {
            if (type == holder.generatorType) {
                holder.btnEdit.setVisibility(params.isEditable() && type == selectedType ? View.VISIBLE : View.INVISIBLE);
                holder.image.setImageBitmap(bitmap);
            }
        });
    }

    @Override
    public int getItemCount() {
        return generatorType.length;
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.generator_type_item_card) View card;
        @BindView(R.id.generator_type_item_text) TextView text;
        @BindView(R.id.generator_type_item_btn_edit) View btnEdit;
        @BindView(R.id.generator_type_item_selected) View isSelected;
        @BindView(R.id.generator_type_item_image) ImageView image;

        GeneratorType generatorType;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void onClick(View v) {
            if (v == btnEdit) {
                listener.onEditSelected();
            } else if (v == card) {
                if (selectedType != generatorType) {
                    selectedType = generatorType;
                    listener.onSelected(selectedType);
                    notifyDataSetChanged();
                } else if (isDeselectAllowed) {
                    selectedType = null;
                    listener.onDeselected();
                    notifyDataSetChanged();
                }
            }
        }
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/view/adapters/GeneratorTypeAdapter.kt
package com.stedi.randomimagegenerator.app.view.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.OvershootInterpolator
import android.widget.ImageView
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.squareup.picasso.Picasso
import com.stedi.randomimagegenerator.app.R
import com.stedi.randomimagegenerator.app.di.ActivityContext
import com.stedi.randomimagegenerator.app.model.data.GeneratorType
import com.stedi.randomimagegenerator.app.other.dim2px
import com.stedi.randomimagegenerator.app.other.nameRes
import com.stedi.randomimagegenerator.app.view.components.RigRequestHandler

class GeneratorTypeAdapter(
        @ActivityContext private val context: Context,
        private val listener: ClickListener,
        private val isDeselectAllowed: Boolean) : RecyclerView.Adapter<GeneratorTypeAdapter.ViewHolder>() {

    private var generatorTypes: Array<GeneratorType> = emptyArray()
    private var selectedType: GeneratorType? = null
    private var targetType: GeneratorType? = null

    private val imageSize = context.dim2px(R.dimen.adapter_rig_image_size)

    interface ClickListener {
        fun onSelected(type: GeneratorType)

        fun onDeselected()

        fun onEditSelected()
    }

    init {
        setHasStableIds(true)
    }

    fun set(generatorTypes: Array<GeneratorType>, selectedType: GeneratorType?, targetType: GeneratorType?) {
        this.generatorTypes = generatorTypes
        this.selectedType = selectedType
        this.targetType = targetType
        notifyDataSetChanged()
    }

    override fun getItemCount() = generatorTypes.size

    override fun getItemId(position: Int) = generatorTypes[position].ordinal.toLong()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GeneratorTypeAdapter.ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.generator_type_item, parent, false))
    }

    override fun onBindViewHolder(holder: GeneratorTypeAdapter.ViewHolder, position: Int) {
        val type = generatorTypes[position]
        holder.generatorType = type

        holder.text.setText(type.nameRes())
        holder.btnEdit.visibility = if (type.isEditable && type === selectedType) View.VISIBLE else View.INVISIBLE
        holder.isSelected.visibility = if (type === selectedType) View.VISIBLE else View.INVISIBLE

        Picasso.get().load(RigRequestHandler.makeFromGenerator(type, targetType, imageSize, imageSize))
                .placeholder(context.resources.getDrawable(R.drawable.ic_texture_rig))
                .into(holder.image)

        holder.card.setOnClickListener(holder)
        holder.btnEdit.setOnClickListener(holder)
    }

    inner class ViewHolder(item: View) : RecyclerView.ViewHolder(item), View.OnClickListener {
        @BindView(R.id.generator_type_item_card) lateinit var card: View
        @BindView(R.id.generator_type_item_text) lateinit var text: TextView
        @BindView(R.id.generator_type_item_btn_edit) lateinit var btnEdit: View
        @BindView(R.id.generator_type_item_selected) lateinit var isSelected: View
        @BindView(R.id.generator_type_item_image) lateinit var image: ImageView

        var generatorType: GeneratorType? = null

        init {
            ButterKnife.bind(this, item)
        }

        override fun onClick(v: View) {
            if (v === btnEdit) {
                listener.onEditSelected()
            } else if (v === card) {
                val generatorType = generatorType ?: return
                if (selectedType !== generatorType) {
                    selectedType = generatorType
                    listener.onSelected(generatorType)
                    runSelectionAnimation()
                    notifyDataSetChanged()
                } else if (isDeselectAllowed) {
                    selectedType = null
                    listener.onDeselected()
                    notifyDataSetChanged()
                }
            }
        }

        private fun runSelectionAnimation() {
            isSelected.scaleX = 0f
            isSelected.scaleY = 0f
            isSelected.alpha = 0f
            btnEdit.alpha = 0f
            isSelected.animate().scaleX(1f).scaleY(1f)
                    .setInterpolator(OvershootInterpolator(1.2f))
                    .alpha(1f).setDuration(200)
                    .withEndAction { btnEdit.animate().alpha(1f).duration = 200 }
        }
    }
}